### PR TITLE
Fixing a performance issue in SSE client

### LINKED ISSUE #198: Performance issue and copyright violation in `_sseclient.py` (for `db.Reference.listen()`)
### [REQUIRED] Step 2: Describe your environment

  * Operating System version: macOS 10.13.6
  * Firebase SDK version: 2.13.0
  * Firebase Product: database

#### Steps to reproduce:

1. Have some Firebase Realtime Database path containing a significant amount of data (>200KB, say).
2. Call `db.reference('/path').listen(callback)`.

As the first update, the server will send the entire contents of that path. On my machine this pegs a CPU core for minutes before the callback is finally called. The issue is the following code: https://github.com/firebase/firebase-admin-python/blob/abd39805f83f906296d5c99b6c4943754154d02c/firebase_admin/_sseclient.py#L93-L96
Since `self.resp_iterator` iterates through the stream one character at a time and `self._event_complete()` runs a regular expression over the entire buffer, this is an O(N^2) operation, which takes a long time when N is large.

**Separately, there is a copyright violation here.** I discovered that `_sseclient.py, despite having a Google copyright header, is largely copied from https://github.com/btubbs/sseclient/blob/master/sseclient.py.

#### Relevant Code:

Run the following code to reproduce:
```python
def listen_issue():
    path = '/sandbox/listen'

    print("Inserting data...")
    start = time.time()
    for i in range(10):
        long_data = ''.join(random.choice(string.ascii_uppercase) for _ in range(20000))
        db.reference(os.path.join(path, str(i))).set(long_data)
    end = time.time()
    print("...done (took %f seconds)" % (end - start))

    done_event = threading.Event()
    listener = db.reference(path).listen(lambda db_event: done_event.set())
    start = time.time()
    print("Waiting for first message from listen()...")
    done_event.wait()  # CPU usage is at 100% here
    end = time.time()
    print("...done (took %f seconds)" % (end - start))
    listener.close()
```

Example output on my machine:
```
Inserting data...
...done (took 1.678544 seconds)
Waiting for first message from listen()...
...done (took 109.180737 seconds)
```

### PR DESCRIPTION
**Pull request by hiranya911**
Improves the streaming RTDB API performance by checking for end of field more efficiently. Specifically, instead of matching a regex on each iteration, we keep track of the last 4 characters seen, and perform a simple string comparison.

Here are some preliminary test results to get an idea of the improvement.

| Node Size  | Time to process (Before) | Time to process (After) |
| ------------- | ------------- | ------------- |
| 100K  | 40.83s  | 1.73s |
| 1M  | - | 11.36s |

Fixes #198

### CODE REVIEW

**Comment by daniel-ziegler** [comment]
Doesn't work for me at all on Mac OS X -- I simply don't receive any listen events. This definitely needs to be tested with a real HTTP stream.

Also note that while this should hopefully improve the constant factor, it's still O(N^2).

**Comment by hiranya911** [comment]
There seems to be a `requests` bug here, which I've reported at https://github.com/requests/requests/issues/4876

The proposed solution essentially reduces N by reading data in chunks. With the fix I've tested payloads (in the unit tests) as large as 10MB, which can be processed under a second.

**Comment by hiranya911** [comment]
Attached is another solution I've been tinkering with. It doesn't change how we read content from `requests` -- just changes how we handle the incoming data:

* Using a list to buffer incoming data as opposed to string concatenation which is inefficient at large sizes.
* Getting rid of the regex check on each iteration. The alternative check I've implemented assumes '\n\n` as the only acceptable terminator, which seems to be correct for Firebase.

With this I was able to read a 1MB node from a production database in under 10 seconds. Feedback welcome.

[sse.txt](https://github.com/firebase/firebase-admin-python/files/2591340/sse.txt)

**Comment by daniel-ziegler** [comment]
I like that solution better since it is indeed linear time. Of course for maximum performance we'd need to figure out how to read in bigger chunks and not process each character individually in Python, but I'm happy with the improvement.

**Comment by hiranya911** [comment]
@daniel-ziegler I've cleaned up the 2nd solution and committed it. Can you give it a try? My performance numbers are added to the PR description.

File: firebase_admin/_sseclient.py [hunk 1]
```diff
@@ -37,6 +40,35 @@ def rebuild_auth(self, prepared_request, response):
...
+
+    def __init__(self):
+        self._buffer = []
+        self._tail = ''
+
+    def append(self, char):
+        self._buffer.append(char)
+        self._tail += char
+        if len(self._tail) > 4:
+            self._tail = self._tail[1:]
```
Comment (daniel-ziegler) [inline firebase_admin/_sseclient.py]: You could also just do `self._tail = self._tail[-4:]` instead of the `if`.

**Comment by daniel-ziegler** [comment]
Seems to work; obviously I haven't tested it very thoroughly.

File: firebase_admin/_sseclient.py (hunk 1 above)
Comment (hiranya911) [reply firebase_admin/_sseclient.py]: Done

File: firebase_admin/_sseclient.py [hunk 2]
```diff
@@ -150,7 +177,7 @@ def parse(cls, raw):
           raw: the raw data to parse.
 
         Returns:
-          Event: newly intialized ``Event`` object with the parameters  initialized.
+          Event: newly intialized ``Event`` object with the parameters initialized.
```
Comment (bklimt) [inline firebase_admin/_sseclient.py]: s/intialized/initialized/

Comment (hiranya911) [reply firebase_admin/_sseclient.py]: Fixed the comment

**Pull request merged**